File: tools/asviz-run.js
#!/usr/bin/env node
//
// Node host for the AllSpeak visualiser framework.
//
// It loads the JS runtime in the bundle order documented in build-allspeak,
// loads the viz plugin, registers the requested source files as the host side of
// the plugin contract, and then runs viz.as exactly as a browser page would.
// Nothing here draws: it is the same "text first" increment, just with a
// command line instead of a page.
//
// Usage:  node tools/asviz-run.js [script.as ...]
//         (default target: codex/en/code/step13.as)

const fs = require('fs');
const path = require('path');
const vm = require('vm');
const { execFileSync } = require('child_process');

const root = path.resolve(__dirname, `..`);
const ANALYSER = path.join(root, `tools`, `asdoc-check.py`);

// The doc-block model comes from the canonical analyser rather than a second parser:
// shelling out keeps its --json output as the single contract between host and
// plugin. A missing analyser just means no prose, and the narrative falls back to the
// anchors instead.
function sectionsFor(target) {
	try {
		return execFileSync(`python3`, [ANALYSER, `--json`, target], { encoding: `utf8` });
	} catch (err) {
		// The analyser exits non-zero when a file has doc-block errors, but it still
		// writes its report to stdout.
		return typeof err.stdout === `string` ? err.stdout : ``;
	}
}

// The runtime expects a browser. Nothing below needs to do anything: compile and
// print never touch the DOM, they only reach for elements that aren't there.
const noop = () => {};

// In a browser, `window` *is* the global object. That matters here: language
// packs are declared with a top-level `var`, and `checkLanguageDirective` finds
// them only as properties of `window`. So model it the same way rather than
// shimming a separate object.
global.window = global;
global.location = { search: `` };
global.localStorage = { getItem: () => null, setItem: noop, removeItem: noop };
global.addEventListener = noop;
global.removeEventListener = noop;

global.document = {
	getElementById: () => null,
	querySelector: () => null,
	createElement: () => ({ style: {}, appendChild: noop, setAttribute: noop }),
	addEventListener: noop,
	body: { appendChild: noop, style: {}, classList: { add: noop, remove: noop } },
	head: { appendChild: noop }
};

// Bundle order from ./build-allspeak (AllSpeak.js is the browser startup hook,
// so it is deliberately omitted — this host starts the framework itself).
const RUNTIME = [
	`Core.js`, `Browser.js`, `MarkdownRenderer.js`, `Webson.js`, `JSON.js`, `MQTT.js`,
	`REST.js`, `Compare.js`, `Condition.js`, `Value.js`, `Run.js`, `Opcodes.js`,
	`Language.js`, `LanguagePack_en.js`, `Compile.js`, `Main.js`
];

const load = (file, from) => {
	vm.runInThisContext(fs.readFileSync(path.join(root, from, file), `utf8`), { filename: file });
};

for (const file of RUNTIME) {
	load(file, `js/allspeak`);
}

// The other language packs. Only the English pack is in the runtime bundle; the
// rest are loaded per page, and a script declaring `language français` cannot be
// compiled without them. This is the multilingual requirement in concrete form.
const packs = [];
for (const file of fs.readdirSync(path.join(root, `js/allspeak`))
	.filter(f => /^LanguagePack_.+\.js$/.test(f)).sort()) {
	if (RUNTIME.includes(file)) continue;
	load(file, `js/allspeak`);
	packs.push(file.replace(/^LanguagePack_/, ``).replace(/\.js$/, ``));
}

// The analysis is only as complete as the domains in scope: a value or condition
// that a missing plugin would have handled cannot be compiled at all. So load the
// plugins the way a host page does — and report which ones made it.
const loaded = [];
const skipped = [];
const pluginDir = path.join(root, `js/plugins`);
for (const file of fs.readdirSync(pluginDir).filter(f => f.endsWith(`.js`)).sort()) {
	try {
		vm.runInThisContext(fs.readFileSync(path.join(pluginDir, file), `utf8`), { filename: file });
		loaded.push(file.replace(/\.js$/, ``));
	} catch (err) {
		skipped.push(`${file} (${err.message.split(`\n`)[0]})`);
	}
}

// Top-level `const` in a vm script lands in the realm's global lexical scope,
// visible to later scripts but not to this module. Hand the two we need across.
vm.runInThisContext(`globalThis.__viz = { AllSpeak, AllSpeak_Viz, AllSpeak_Language, ` +
	`AllSpeak_LanguagePack_en, AllSpeak_Run };`);
const {
	AllSpeak, AllSpeak_Viz, AllSpeak_Language, AllSpeak_LanguagePack_en, AllSpeak_Run
} = globalThis.__viz;

AllSpeak_Language.init(AllSpeak_LanguagePack_en);
AllSpeak.timestamp = Date.now();
// The rest of what AllSpeak_Startup does in a browser before any script runs.
AllSpeak.scripts = {};

process.stderr.write(`domains: ${Object.keys(AllSpeak.domain).join(`, `)}\n`);
process.stderr.write(`plugins loaded: ${loaded.join(`, `)}\n`);
process.stderr.write(`language packs: en, ${packs.join(`, `)}\n`);
for (const note of skipped) {
	process.stderr.write(`plugin skipped: ${note}\n`);
}

// The host side of the plugin contract: which script is being looked at, and its
// source text. A browser host would fill these from the editor buffer instead.
const targets = process.argv.slice(2);
if (targets.length === 0) {
	targets.push(`codex/en/code/step13.as`);
}

const framework = fs.readFileSync(path.join(root, `viz.as`), `utf8`);

// A failing target is a finding, not a reason to stop: keep going so a whole
// corpus can be swept in one pass, and report each failure against its name.
let failures = 0;
const firstLine = (err) => String((err && err.message) || err).split(`\n`)[0];

for (const target of targets) {
	if (!fs.existsSync(path.join(root, target))) {
		process.stderr.write(`asviz-run: no such file: ${target}\n`);
		failures++;
		continue;
	}
	AllSpeak_Viz.target = target;
	AllSpeak_Viz.sources[target] = fs.readFileSync(path.join(root, target), `utf8`);
	AllSpeak_Viz.sections[target] = sectionsFor(path.join(root, target));
	AllSpeak_Viz.problems = [];

	// Compile the framework ourselves rather than via AllSpeak.start: start() is
	// once-only and routes errors through reportError, which is built for a page
	// with a compiler instance still in scope.
	let program;
	try {
		const source = AllSpeak.tokeniseFile(framework.split(`\n`));
		program = AllSpeak.compileScript(source, null, null, null);
	} catch (err) {
		process.stderr.write(`FAIL ${target}: framework compile: ${firstLine(err)}\n`);
		failures++;
		continue;
	}
	// The framework's own `script Viz` line registers it under that name; give it
	// a numeric id instead, so the name registry stays clean across targets.
	delete AllSpeak.scripts[program.script];
	program.script = AllSpeak.scriptIndex++;
	AllSpeak.scripts[program.script] = program;
	program.running = true;
	try {
		AllSpeak_Run.run(program, 0);
	} catch (err) {
		process.stderr.write(`FAIL ${target}: ${firstLine(err)}\n`);
		failures++;
	}
	if (AllSpeak_Viz.problems.length > 0) {
		process.stderr.write(`FAIL ${target}: ${AllSpeak_Viz.problems[0]}\n`);
		failures++;
	}
}

if (failures > 0) {
	process.stderr.write(`${failures} of ${targets.length} target(s) failed\n`);
	process.exitCode = 1;
}

File: tools/asviz-run.py
#!/usr/bin/env python3
"""Node-host counterpart for the Python runtime: run viz.as against a script.

The framework (viz.as) is the same file the JS side uses. What differs is the
host: this one registers the plugin domain and points it at a target, exactly as
a browser page would load the plugin and hand it the editor buffer.

Usage:  python3 tools/asviz-run.py <script.as> [...]
        python3 tools/asviz-run.py --run <script.as> [...]   (also run it, and report
                                                              the recording its markers made)
        python3 tools/asviz-run.py --run --trace=<file.json> <script.as>
                                                             (also write the recording as a
                                                              Chrome trace; load it at
                                                              ui.perfetto.dev)
        python3 tools/asviz-run.py --run --trace=<file.json> --trace-pretty <script.as>
                                                             (indent it for reading; the
                                                              default is compact, since a
                                                              trace is mostly read by a
                                                              viewer)

Paths are relative to the current directory, which is where `Program` resolves
them from too, so run it from the repository root.
"""

import json
import os
import subprocess
import sys

USAGE = """usage: asviz-run.py [--run|-r] [--trace=<file.json>] [--trace-pretty[=<file.json>]]
                       [--trace-compact[=<file.json>]] <script.as> [...]
       A trace needs --run, and one target. Either trace flag may carry the path, so
       --trace-pretty=pretrace.json is the same as --trace=pretrace.json --trace-pretty.
"""

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, 'allspeak-py'))
sys.path.insert(0, os.path.join(ROOT, 'allspeak-py', 'plugins'))

from allspeak import Program          # noqa: E402
import as_viz                         # noqa: E402
from as_viz import Viz, VizState, Recorder      # noqa: E402

FRAMEWORK = os.path.join(ROOT, 'viz.as')
ANALYSER = os.path.join(ROOT, 'tools', 'asdoc-check.py')


def sectionsFor(target):
    """The doc-block model, from the canonical analyser rather than a second parser.

    Shelling out keeps the analyser's --json output as the single contract between
    the two: the host acquires, the plugin interprets. A missing analyser just means
    no prose, and the narrative falls back to the anchors.
    """
    try:
        done = subprocess.run([sys.executable, ANALYSER, '--json', target],
                              capture_output=True, text=True, timeout=60)
        return done.stdout
    except (OSError, subprocess.SubprocessError):
        return ''


def runTarget(target):
    """Run the instrumented script so the markers have something to record.

    This executes the target, side effects and all. That is the point of a recording:
    the data cannot be collected without the work being done, so the tool is meant to be
    used with care rather than fenced in.
    """
    program = Program(target, testMode=True)
    program.summaryPrinted = True
    # The markers are core syntax now, so the target needs no plugin domain to compile.
    # The recorder is the facility that makes a run a *recording*: it watches the markers
    # go past and collects the trace. Attached here rather than by a keyword, because
    # collecting data is not something the script should have to ask for.
    program.recorder = Recorder()
    VizState.trace[program.scriptName] = program.recorder
    program.start()
    # The run is over, so a window still open ends here rather than whenever a reader looks.
    program.recorder.finish()
    return program.recorder


def main(argv):
    argv = list(argv)
    run = False
    trace = None
    pretty = False
    while argv and argv[0].startswith('-'):
        flag = argv.pop(0)
        if flag in ('--run', '-r'):
            run = True
        elif flag.startswith('--trace='):
            trace = flag.split('=', 1)[1]
        elif flag == '--trace-pretty':
            pretty = True
        elif flag == '--trace-compact':
            pass                       # the default, accepted so both spellings work
        elif flag.startswith('--trace-pretty='):
            pretty = True
            trace = flag.split('=', 1)[1]
        elif flag.startswith('--trace-compact='):
            trace = flag.split('=', 1)[1]
        elif flag == '--trace':
            # `--trace` alone would have to guess a filename, and guessing writes files
            # nobody asked for. The equals form is unambiguous beside a list of targets.
            sys.stderr.write('asviz-run: --trace needs a path: --trace=<file.json>\n')
            return 1
        else:
            # A mistyped flag is the likeliest reason to land here, so say what the
            # flags are rather than only what this one is not.
            sys.stderr.write(f'asviz-run: unknown option: {flag}\n')
            sys.stderr.write(USAGE)
            return 1
    targets = argv or ['codex/en/code/step13.as']
    if trace and not run:
        sys.stderr.write('asviz-run: a trace records a run, so it needs --run too\n')
        return 1
    if trace and len(targets) > 1:
        sys.stderr.write('asviz-run: --trace takes one target, not several\n')
        return 1
    for target in targets:
        if not os.path.exists(target):
            sys.stderr.write(f'asviz-run: no such file: {target}\n')
            return 1

    with open(FRAMEWORK, 'r', encoding='utf-8') as f:
        framework = f.read()

    failures = 0
    for target in targets:
        # The host side of the contract: which script, and its text. Unlike the JS
        # host this one need not supply the text — the plugin can read a file — but
        # supplying it keeps the two hosts comparable.
        VizState.target = target
        with open(target, 'r', encoding='utf-8') as f:
            VizState.sources[target] = f.read()
        VizState.sections[target] = sectionsFor(target)
        VizState.problems = []

        if run:
            sys.stderr.write(f'asviz-run: running {target}\n')
            recorder = runTarget(target)
            # Written before the framework runs so a trace survives a framework failure,
            # and because a recording is a fact about the run, not about the report.
            if trace:
                windows = recorder.finishedWindows()
                document = as_viz.traceDocument(target, windows)
                try:
                    with open(trace, 'w', encoding='utf-8') as f:
                        # Compact by default: whitespace does not matter to a JSON reader,
                        # and a trace's first consumer is a viewer, not a person. Nothing is
                        # lost when someone does want to read one, since a compact file can
                        # be indented without re-running: python3 -m json.tool <file>
                        if pretty:
                            json.dump(document, f, indent=2)
                        else:
                            json.dump(document, f, separators=(',', ':'))
                except OSError as e:
                    sys.stderr.write(f'asviz-run: cannot write {trace}: {e}\n')
                    return 1
                sys.stderr.write(
                    f'asviz-run: trace: {trace} '
                    f'({len(document["traceEvents"])} events, '
                    f'{len(windows)} window(s))\n')

        program = Program(FRAMEWORK, testMode=True)
        program.useClass(Viz)          # the domain must exist before viz.as compiles
        # testMode keeps the compile quiet; this keeps its end-of-run test summary
        # quiet too, so the only thing on stdout is the model.
        program.summaryPrinted = True
        try:
            program.start()
        except BaseException as e:     # noqa: BLE001 - report per target, keep going
            # AllSpeak's RuntimeError does not stringify, so fall back to the type
            # name and point at the log the runtime already printed.
            detail = str(e) or f'{type(e).__name__} (see the log above)'
            sys.stderr.write(f'FAIL {target}: {detail}\n')
            failures += 1
            continue
        if VizState.problems:
            failures += 1
            sys.stderr.write(f'FAIL {target}: {VizState.problems[0]}\n')

    if failures:
        sys.stderr.write(f'{failures} of {len(targets)} target(s) failed\n')
    return 1 if failures else 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
